File: c++/2024/2024-001/part-001/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <cstdlib>  // for std::abs

// Function to calculate the total distance between two sorted lists
void calculate_total_distance(const std::string &file_path) {
    std::ifstream file(file_path);
    if (!file) {
        std::cerr << "Could not open the file: " << file_path << std::endl;
        return;
    }

    std::vector<int> left_list;
    std::vector<int> right_list;
    int left, right;

    // Read the file and load the location IDs into two vectors
    while (file >> left >> right) {
        left_list.push_back(left);
        right_list.push_back(right);
    }
    file.close();

    // Sort both lists
    std::sort(left_list.begin(), left_list.end());
    std::sort(right_list.begin(), right_list.end());

    // Calculate the total distance
    int total_distance = 0;
    for (size_t i = 0; i < left_list.size(); ++i) {
        total_distance += std::abs(left_list[i] - right_list[i]);
    }

    // Output the total distance
    std::cout << "Total Distance: " << total_distance << std::endl;
}

int main() {
    // File path for input data
    const std::string file_path = "input.txt";

    // Call the function to calculate the total distance
    calculate_total_distance(file_path);

    return 0;
}

File: Swift/2024/2024-001/part-001/main.swift
import Foundation

// Function to calculate total distance
func calculateTotalDistance(filePath: String) -> Int {
    // Read the file content
    guard let fileContent = try? String(contentsOfFile: filePath) else {
        print("Failed to read file at \(filePath)")
        return 0
    }
    
    // Split file content into lines
    let lines = fileContent.split(separator: "\n")
    
    // Initialize two lists
    var leftList: [Int] = []
    var rightList: [Int] = []
    
    // Parse the lines into two lists
    for line in lines {
        let parts = line.split(separator: " ").compactMap { Int($0) }
        if parts.count == 2 {
            leftList.append(parts[0])
            rightList.append(parts[1])
        }
    }
    
    // Sort both lists
    leftList.sort()
    rightList.sort()
    
    // Calculate the total distance
    let totalDistance = zip(leftList, rightList).reduce(0) { $0 + abs($1.0 - $1.1) }
    
    return totalDistance
}

// Input file path
let filePath = "input.txt"

// Calculate and print the total distance
let totalDistance = calculateTotalDistance(filePath: filePath)
print("Total Distance: \(totalDistance)")
